capa/features/extractors/binexport2/insn.py

```diff
@@ -0,0 +1,414 @@
+# Copyright (C) 2023 Mandiant, Inc. All Rights Reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at: [package root]/LICENSE.txt
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+#  is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+import io
+import logging
+from typing import List, Tuple, Iterator
+
+import pefile
+from elftools.elf.elffile import ELFFile
+
+import capa.features.extractors.helpers
+import capa.features.extractors.strings
+from capa.features.insn import API, Number, Mnemonic, OperandNumber
+from capa.features.common import Bytes, String, Feature, Characteristic
+from capa.features.address import Address, AbsoluteVirtualAddress
+from capa.features.extractors.binexport2 import AnalysisContext, FunctionContext, InstructionContext
+from capa.features.extractors.base_extractor import BBHandle, InsnHandle, FunctionHandle
+from capa.features.extractors.binexport2.binexport2_pb2 import BinExport2
+
+logger = logging.getLogger(__name__)
+
+
+def extract_insn_api_features(fh: FunctionHandle, _bbh: BBHandle, ih: InsnHandle) -> Iterator[Tuple[Feature, Address]]:
+    fhi: FunctionContext = fh.inner
+    ii: InstructionContext = ih.inner
+
+    be2 = fhi.ctx.be2
+    idx = fhi.ctx.idx
+    analysis = fhi.ctx.analysis
+
+    instruction = be2.instruction[ii.instruction_index]
+
+    if not instruction.call_target:
+        return
+
+    for call_target_address in instruction.call_target:
```

Comment:
We can use Context.Vertex.Type.THUNK and Context.Vertex.Type.IMPORTED

Reply:
Thunk chains should be fully linked by the Ghidra BinExport exporter as of https://github.com/google/binexport/commit/031e5c3d64f33ad99483394552b14e5387a9bdff.

Reply:
I'll be pushing code for this soon. Just a couple of items to recheck.

Reply:
I wonder if we should pre-compute all the thunks and their targets and put them in the `BinExport2Analysis` rather than walking the chain each time we encounter a thunk call. 

Two benefits:

  1. maybe a minor performance enhancement, and
  2. keep the insn feature handling simpler

Reply:
This makes sense. I'll take a stab at it quick. This may also be something to consider for capa's other extractors

Reply:
See https://github.com/mandiant/capa/commit/9a9d5a2e33ea2363ba59f030688ce03c434da598